igvm/commands.py

```diff
@@ -51,6 +51,45 @@ def _check_defined(vm, fail_hard=True):
             log.info(error)
 
 
+@with_fabric_settings
+def evacuate(hv_hostname, offline=None, dry_run=False):
+    """Move all VMs out of a hypervisor
+
+    Move all VMs out of a hypervisor and put it to state maintenance.
+
+    Offline can be passed without arguments or with a list strings matching
+    function attributes. If just passed all VMs will be migrated offline. If
+    a list of strings is passed only those matching will be migrate offline.
+    """
+
+    hv = Query({'hostname': hv_hostname}, ['state'])
+    if len(hv) == 0:
+        print('No such hypervisor {}'.format(hv_hostname))
+        exit(1)
+
+    if dry_run:
+        print('Setting {} to state maintenance'.format(hv_hostname))
```

Comment:
This should be done only after HV is `igvm_locked`.

Reply:
I adjusted it. Can you please check if I used the context manager for locking hv correctly ?

Reply:
I had a quick glance at it and think it should work. That said using an exit stack is not required if you only want to acquire one context manager. Therefore you can replace

```python3
    with ExitStack() as es:
        hv = es.enter_context(
            _get_hypervisor(hv_hostname, ignore_reserved=True)) 
```

with the following to the same end:
```python3
    with _get_hypervisor(hv_hostname, ignore_reserved=True) as hv:
        [...]
```

I wrote up this little piece about exit stacks and context managers after Emres presentation to be able look that stuff up again. Maybe it's helpful to you too :)
https://hackmd.io/ys74lBqQS5SuxIBMEHIH_Q#ExitStack